From fd0e7030bd20927481ff2250b1b82d58d002eb3e Mon Sep 17 00:00:00 2001
From: Dylan Baker <dylan@pnwbakers.com>
Date: Wed, 26 Feb 2025 09:23:12 -0800
Subject: [PATCH 1/4] modules/rust: Update bindgen target error checking for
 bindgen >= 0.71

Which has both changed the error message and relaxed the check. In
theory we wouldn't hit this unless you have a very old bindgen and a
very new rustc.
---
 mesonbuild/modules/rust.py | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/mesonbuild/modules/rust.py b/mesonbuild/modules/rust.py
index 3638964f28ef..34a3f68d8aed 100644
--- a/mesonbuild/modules/rust.py
+++ b/mesonbuild/modules/rust.py
@@ -1,5 +1,5 @@
 # SPDX-License-Identifier: Apache-2.0
-# Copyright © 2020-2024 Intel Corporation
+# Copyright © 2020-2025 Intel Corporation
 
 from __future__ import annotations
 import itertools
@@ -259,9 +259,11 @@ def bindgen(self, state: ModuleState, args: T.List, kwargs: FuncBindgen) -> Modu
                 _, _, err = mesonlib.Popen_safe(
                     T.cast('T.List[str]', self._bindgen_bin.get_command()) +
                     ['--rust-target', self._bindgen_rust_target])
-                # Sometimes this is "invalid Rust target" and sometimes "invalid
-                # rust target"
-                if 'Got an invalid' in err:
+                # < 0.71: Sometimes this is "invalid Rust target" and
+                # sometimes "invalid # rust target"
+                # >= 0.71: error: invalid value '...' for '--rust-target <RUST_TARGET>': "..." is not a valid Rust target, accepted values are of the form ...
+                # It's also much harder to hit this in 0.71 than in previous versions
+                if 'Got an invalid' in err or 'is not a valid Rust target' in err:
                     self._bindgen_rust_target = None
 
         name: str

From f07bcb98168941d69cda805e5660313a3c4891b2 Mon Sep 17 00:00:00 2001
From: Dylan Baker <dylan@pnwbakers.com>
Date: Wed, 26 Feb 2025 09:48:52 -0800
Subject: [PATCH 2/4] modules/rust: set --rust-edition for bindgen >= 0.71

But only if the user hasn't manually set it.
---
 mesonbuild/modules/rust.py | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/mesonbuild/modules/rust.py b/mesonbuild/modules/rust.py
index 34a3f68d8aed..0891243eb7c7 100644
--- a/mesonbuild/modules/rust.py
+++ b/mesonbuild/modules/rust.py
@@ -66,6 +66,7 @@ def __init__(self, interpreter: Interpreter) -> None:
             self._bindgen_rust_target: T.Optional[str] = interpreter.compilers.host['rust'].version
         else:
             self._bindgen_rust_target = None
+        self._bindgen_set_std = False
         self.methods.update({
             'test': self.test,
             'bindgen': self.bindgen,
@@ -266,6 +267,10 @@ def bindgen(self, state: ModuleState, args: T.List, kwargs: FuncBindgen) -> Modu
                 if 'Got an invalid' in err or 'is not a valid Rust target' in err:
                     self._bindgen_rust_target = None
 
+            # TODO: Executable needs to learn about get_version
+            if isinstance(self._bindgen_bin, ExternalProgram):
+                self._bindgen_set_std = mesonlib.version_compare(self._bindgen_bin.get_version(), '>= 0.71')
+
         name: str
         if isinstance(header, File):
             name = header.fname
@@ -335,6 +340,11 @@ def bindgen(self, state: ModuleState, args: T.List, kwargs: FuncBindgen) -> Modu
             kwargs['args'] + inline_wrapper_args
         if self._bindgen_rust_target and '--rust-target' not in cmd:
             cmd.extend(['--rust-target', self._bindgen_rust_target])
+        if self._bindgen_set_std and '--rust-edition' not in cmd:
+            rust_std = state.environment.coredata.optstore.get_value('rust_std')
+            assert isinstance(rust_std, str), 'for mypy'
+            if rust_std != 'none':
+                cmd.extend(['--rust-edition', rust_std])
         cmd.append('--')
         cmd.extend(kwargs['c_args'])
         cmd.extend(clang_args)

From c979f0cfcaf63d57855c07799edc2e92c4b9fb46 Mon Sep 17 00:00:00 2001
From: Dylan Baker <dylan@pnwbakers.com>
Date: Wed, 26 Feb 2025 09:34:44 -0800
Subject: [PATCH 3/4] compilers/detect: Split -beta and -nightly suffixes from
 rustc

Store both a full version with the nightly and beta suffixes, and the
version as just X.Y.Z. This sort of distinction is why full_version
exists.

This fixes an issue with the bindgen module where we pass invalid
version of X.Y.Z-beta and X.Y.Z-nightly.
---
 mesonbuild/compilers/detect.py | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/mesonbuild/compilers/detect.py b/mesonbuild/compilers/detect.py
index d4ad4badeefd..784e00d3dc76 100644
--- a/mesonbuild/compilers/detect.py
+++ b/mesonbuild/compilers/detect.py
@@ -1047,7 +1047,11 @@ def detect_rust_compiler(env: 'Environment', for_machine: MachineChoice) -> Rust
             popen_exceptions[join_args(compiler + arg)] = e
             continue
 
-        version = search_version(out)
+        # Full version contains the "-nightly" or "-beta" suffixes, but version
+        # should just be X.Y.Z
+        full_version = search_version(out)
+        version = full_version.split('-', 1)[0]
+
         cls: T.Type[RustCompiler] = rust.RustCompiler
 
         # Clippy is a wrapper around rustc, but it doesn't have rustc in its
@@ -1063,7 +1067,8 @@ def detect_rust_compiler(env: 'Environment', for_machine: MachineChoice) -> Rust
             except OSError as e:
                 popen_exceptions[join_args(compiler + arg)] = e
                 continue
-            version = search_version(out)
+            full_version = search_version(out)
+            version = full_version.split('-', 1)[0]
 
             cls = rust.ClippyRustCompiler
             mlog.deprecation(
@@ -1143,7 +1148,7 @@ def detect_rust_compiler(env: 'Environment', for_machine: MachineChoice) -> Rust
             env.coredata.add_lang_args(cls.language, cls, for_machine, env)
             return cls(
                 compiler, version, for_machine, is_cross, info,
-                linker=linker)
+                linker=linker, full_version=full_version)
 
     _handle_exceptions(popen_exceptions, compilers)
     raise EnvironmentException('Unreachable code (exception to make mypy happy)')

From d939662b4a836cded18610eee3d13799b28c885f Mon Sep 17 00:00:00 2001
From: Dylan Baker <dylan@pnwbakers.com>
Date: Wed, 26 Feb 2025 09:39:24 -0800
Subject: [PATCH 4/4] modules/rust: use 'nightly' as the bindgen version if the
 compiler is nightly

This adds tracking for both nightly and beta to the rust compiler.
---
 mesonbuild/compilers/rust.py | 2 ++
 mesonbuild/modules/rust.py   | 5 ++++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/mesonbuild/compilers/rust.py b/mesonbuild/compilers/rust.py
index 3acc30e5458c..475b7a402aa8 100644
--- a/mesonbuild/compilers/rust.py
+++ b/mesonbuild/compilers/rust.py
@@ -100,6 +100,8 @@ def __init__(self, exelist: T.List[str], version: str, for_machine: MachineChoic
         if 'link' in self.linker.id:
             self.base_options.add(OptionKey('b_vscrt'))
         self.native_static_libs: T.List[str] = []
+        self.is_beta = '-beta' in full_version
+        self.is_nightly = '-nightly' in full_version
 
     def needs_static_linker(self) -> bool:
         return False
diff --git a/mesonbuild/modules/rust.py b/mesonbuild/modules/rust.py
index 0891243eb7c7..c455ea227996 100644
--- a/mesonbuild/modules/rust.py
+++ b/mesonbuild/modules/rust.py
@@ -24,6 +24,7 @@
 if T.TYPE_CHECKING:
     from . import ModuleState
     from ..build import IncludeDirs, LibTypes
+    from ..compilers.rust import RustCompiler
     from ..dependencies import Dependency, ExternalLibrary
     from ..interpreter import Interpreter
     from ..interpreter import kwargs as _kwargs
@@ -58,12 +59,14 @@ class RustModule(ExtensionModule):
     """A module that holds helper functions for rust."""
 
     INFO = ModuleInfo('rust', '0.57.0', stabilized='1.0.0')
+    _bindgen_rust_target: T.Optional[str]
 
     def __init__(self, interpreter: Interpreter) -> None:
         super().__init__(interpreter)
         self._bindgen_bin: T.Optional[T.Union[ExternalProgram, Executable, OverrideProgram]] = None
         if 'rust' in interpreter.compilers.host:
-            self._bindgen_rust_target: T.Optional[str] = interpreter.compilers.host['rust'].version
+            rustc = T.cast('RustCompiler', interpreter.compilers.host['rust'])
+            self._bindgen_rust_target = 'nightly' if rustc.is_nightly else rustc.version
         else:
             self._bindgen_rust_target = None
         self._bindgen_set_std = False
